src/pudl/helpers.py

```diff
@@ -2185,3 +2194,219 @@ def retry(
+    US numbers: ###-###-####
+    International numbers with the international code at the beginning.
+    Numbers with extensions will be appended with "x#".
+    Non-numeric entries will be returned as np.nan. Entries with fewer than
+    10 digits will be returned with no hyphens.
+
+    Args:
+    df: The DataFrame to modify.
+    columns: A list of the names of the columns that need to be standardized
+    Returns:
+    The modified DataFrame with standardized phone numbers in the same column.
+    """
+
+    # Function to clean and standardize phone numbers
+    def standardize_phone_number(phone: str) -> str:
+        # Check if the value is NaN, return it as is
+        if pd.isna(phone):
+            return phone
+
+        # Split phone number and extension, if present
+        phone_parts = re.split(r"[xX]", str(phone))
+        phone_main = phone_parts[0]  # Main phone number
+        extension = (
+            phone_parts[1].strip() if len(phone_parts) > 1 else None
+        )  # Extension, if it exists
+
+        # Remove unwanted characters (parentheses, spaces, periods, and dashes) from the main phone number
+        phone_main = re.sub(r"[^\d]", "", phone_main)  # Keep only digits
+
+        # Grab the length for the next series of formatting steps
+        phone_main_len = len(phone_main)
+
+        # If phone_main is all zeroes, return np.nan
+        if phone_main == "0" * phone_main_len:
+            return np.nan
+
+        # If the phone number has exactly 10 digits, format it as XXX-XXX-XXXX
+        if phone_main_len == 10:
+            formatted_phone = f"{phone_main[:3]}-{phone_main[3:6]}-{phone_main[6:]}"
+        # Return all numbers that are not 10 digits as-is
+        else:
+            formatted_phone = phone_main
+
+        # Add the extension back if present
+        if extension:
+            return f"{formatted_phone}x{extension}"
+
+        return formatted_phone
+
+    # Apply the standardization function directly to the specified columns
+    for column in columns:
+        df[column] = df[column].apply(standardize_phone_number)
+
+    return df
+
+
+def analyze_missing_values(df: pd.DataFrame) -> list[str]:
+    """Analyze columns of a DataFrame for missing or invalid values. Note that this is purely for analysis
+    and does not perform any data transformation or cleaning.
+    This function checks each column for missing or custom missing values and prints
+    a summary of the findings for string (object), numeric, and datetime columns.
+
+    Args:
+        df: The DataFrame to analyze.
+
+    Returns:
+        exception_cols: List of names of columns that couldn't be analyzed due to a caught exception.
+    """
+    nan_cols = []
+    exception_cols = []
+
+    # Define custom missing value markers
+    custom_missing_values = [
+        "",
+        " ",
+        "NA",
+        "N/A",
+        "NULL",
+        "-",
+        "None",
+        "NaN",
+        "?",
+        "*",
+        "#",
+    ]
```

Comment:
If this function is going to stick around post-PR then we probably want the user to be able to define their own set of custom missing values -- though these could be retained as a default set.

Reply:
Probably also want to switch all of the `print()` statements to `logger.info()` outputs so it can be managed along with all of the other outputs.

Reply:
Updated